### ISSUE TITLE
redis pool error when PING is sent

### ISSUE THREAD

**Issue by tdyas** [bikeshedder/deadpool#97]
I am using the `deadpool` / `deadpool-redis` crates at `8134ba16896a9149301ee9e2945744ba8210dad3` with `redis` v0.20.1. My backend Redis is AWS Elasticache for Redis with redis6 profile. Rust is v1.52.1.

I am seeing the following error messages when I deployed my service to production:

```
Response was of incompatible type: "Response status not valid boolean" (response was status ("PONG"))
```

Looking at the recycle code at https://github.com/bikeshedder/deadpool/blob/bcbe5677dd9aa836eb32787f5d25bf4824460731/redis/src/lib.rs#L230-L236, this is the only instance of a Redis PING command across my service and these crates, but does not seem obvious why an attempt would be made to decode as a `bool` there.

I also saw other error messages from my service about incorrect results being returned which suggests something was possibly out-of-sync with the connection, but that will need further investigation.

**Comment by tdyas** [comment]
My service has places that it tries to decode an EXISTS as a `bool`, which would raise the question why a PONG response was being returned at that locations.

**Comment by bikeshedder** [comment]
That `PING` command is used to check the health of a connection. It sounds like some part of the `redis` crate is not drop safe. The only way I can explain the behavior is like that:

- `query_async` is used but not `.await`-ed until the response is available.
- The `Client` is returned to the pool.
- The response becomes available.
- Upon recycling a `PING` query is issued. That code doesn't check if the response is a `PONG` but just checks if the query succeeds. Therefore consuming the response from the "aborted" query.
- The next query recieves the `PONG` from the recycling and all queries after that receive the response from the query right before it.

I was able to reproduce this issue using `futures::FutureExt::now_or_never`:
```rust
#[tokio::test]
async fn test_aborted_command() {
    use deadpool_redis::redis::AsyncCommands;
    let pool = create_pool();
    {
        let mut conn = pool.get().await.unwrap();
        let _: () = conn.set("deadpool/abort_test_wrong", "wrong").await.unwrap();
        let _: () = conn.set("deadpool/abort_test_right", "right").await.unwrap();
    }
    {
        let mut conn = pool.get().await.unwrap();
        let fut = conn.get::<_, ()>("deadpool/abort_test_wrong");
        // Poll the future once. This does execute the query but does not
        // wait for the response. The connection now has a request queued
        // and the response to it will be returned when using the connection
        // the next time.
        fut.now_or_never();
    }
    {
        let mut conn = pool.get().await.unwrap();
        let value: String = conn.get("deadpool/abort_test_right").await.unwrap();
        assert_eq!(value, "right");
    }
}

```

I have a workaround for this issue available and will release a new version of `deadpool-redis` asap.

**Comment by bikeshedder** [comment]
A simpler reproduction:
```rust
#[tokio::test]
async fn test_aborted_command() {
    let pool = create_pool();
    {
        let mut conn = pool.get().await.unwrap();
        // Poll the future once. This does execute the query but does not
        // wait for the response. The connection now has a request queued
        // and the response to it will be returned when using the connection
        // the next time.
        cmd("PING")
            .arg("wrong")
            .query_async::<_, String>(&mut conn)
            .now_or_never();
    }
    {
        let mut conn = pool.get().await.unwrap();
        let value: String = cmd("PING")
            .arg("right")
            .query_async(&mut conn)
            .await
            .unwrap();
        assert_eq!(value, "right");
    }
}
```

**Comment by tdyas** [comment]
> query_async is used but not .await-ed until the response is available.

My service exposes a gRPC streaming interface to its own clients. Those clients can close that stream early for various reasons (including network issues). Thus, the Future in which the RPC call occurs can get dropped (and that is what would be polling the Redis Future).

**Comment by bikeshedder** [comment]
I just reported the `redis` issue: mitsuhiko/redis-rs#489

I just completed a workaround for this in `deadpool-redis`. I'm preparing a `deadpool-redis 0.8.0` release right now.

**Comment by bikeshedder** [comment]
@tdyas I just released `deadpool-redis 0.8.0`. Could you please give it a try and report if that fixes your problems?

- https://crates.io/crates/deadpool-redis/0.8.0

This release includes more than just this fix so make sure to check the CHANGELOGs:
- https://github.com/bikeshedder/deadpool/blob/master/redis/CHANGELOG.md
- https://github.com/bikeshedder/deadpool/blob/master/CHANGELOG.md

The API itself didn't change much. The main difference is the possibility to select a runtime to be used so `deadpool` is no longer limited to timeouts of the `tokio` runtime.

**Comment by tdyas** [comment]
> Could you please give it a try and report if that fixes your problems?

I will be able to do so next week. Thanks!

**Comment by bikeshedder** [comment]
I just released deadpool-redis 0.6.2 and deadpool-redis 0.7.2 which both include this fix:

- https://crates.io/crates/deadpool-redis/0.6.2
- https://crates.io/crates/deadpool-redis/0.7.2

See #98
